### PR TITLE
Task 4

### PR DESCRIPTION
**Pull request by Baravar**
**Task 4.1** - DONE

- [x] Use AWS Console to create two database tables in DynamoDB. Expected schemas for products and stocks:
- [x] Write a script to fill tables with test examples. Store it in your Github repository. Execute it for your DB to fill data.

**Task 4.2** - DONE

- [x] Extend your serverless.yml file with data about your database table and pass it to lambda’s environment variables section.
- [x] Integrate the getProductsList lambda to return via GET /products request a list of products from the database (joined stocks and products tables).
- [x] Implement a Product model on FE side as a joined model of product and stock by productId.
- [x] FE: One product model as a result of BE models join (product and it's stock)
- [x] Integrate the getProductsById lambda to return via GET /products/{productId} request a single product from the database.

**Task 4.3** - DONE
- [x] Create a lambda function called createProduct under the same Serverless config file (i.e. serverless.yaml) of Product Service which will be triggered by the HTTP POST method.
The requested URL should be /products.
Implement its logic so it will be creating a new item in a Products table.
- [x] Save the URL (API Gateway URL) to execute the implemented lambda functions for later - you'll need to provide it in the PR (e.g in PR's description) when submitting the task.
**endpoint:**
POST - https://4ipnog1d7l.execute-api.us-east-1.amazonaws.com/products
Valid Payload Sample: 
`{ "count":7, "size":[ "S", "M", "XL" ], "description":"Home jersey", "price":85, "title":"Man City - Home Kit 22-23", "image":"https://shop.mancity.com/dw/image/v2/BDWJ_PRD/on/demandware.static/-/Sites-master-catalog-MAN/default/dw96c87297/images/large/701221508001_pp_01_mcfc.png?sw=1600&sh=1600&sm=fit" }`
Invalid Payload Sample: 
`{ "countzzz":7, "size":[ "S", "M", "XL" ], "description":"Home jersey", "price":85, "title":"Man City - Home Kit 22-23", "image":"https://shop.mancity.com/dw/image/v2/BDWJ_PRD/on/demandware.static/-/Sites-master-catalog-MAN/default/dw96c87297/images/large/701221508001_pp_01_mcfc.png?sw=1600&sh=1600&sm=fit" }`

**Evaluation criteria (70 points for covering all criteria)**
Reviewers should verify the lambda functions by invoking them through provided URLs:
GET All Products: https://4ipnog1d7l.execute-api.us-east-1.amazonaws.com/products
GET Product: https://4ipnog1d7l.execute-api.us-east-1.amazonaws.com/products/acea20eb-e375-4310-a68d-fd11fdbb0a54
GET Product (failing): https://4ipnog1d7l.execute-api.us-east-1.amazonaws.com/products/acea20eb-e375-4310-a68d-fd11fdbb0a
POST Product - https://4ipnog1d7l.execute-api.us-east-1.amazonaws.com/products (find payload sample above)

- [x] Task 4.1 is implemented
- [x] Task 4.2 is implemented lambda links are provided and returns data
- [x] Task 4.3 is implemented lambda links are provided and products is stored in DB (call Task 4.2 to see the product)
Your own Frontend application is integrated with Product Service (/products API) and products from Product Service are represented on Frontend. Link to a working Frontend application is provided for cross-check reviewer.
**UI:** https://d3lc6lhdpvhm1c.cloudfront.net/ 

Additional (optional) tasks
- [x] +6 (All languages) - POST /products lambda functions returns error 400 status code if product data is invalid
- [x] +6 (All languages) - All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
- [x] +6 (All languages) - All lambdas do console.log for each incoming requests and their arguments

Not Done, but DynamoDB refs. are managed using AWS System Manager and Env. Variables (deserves +2 points 😄 ):
- [ ] +6 (All languages) - Use RDS instance instead fo DynamoDB tables. Do not commit your environment variables in serverless.yml to github!
- [ ] +6 (All languages) - Transaction based creation of product (in case stock creation is failed then related to this stock product is not created and not ready to be used by the end user and vice versa) (https://devcenter.kinvey.com/nodejs/tutorials/bl-transactional-support, https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/transaction-apis.html)

### CODE REVIEW

File: products-service/services/dynamodb/products-data.service.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+   
+   const productsData = Promise.all([products, productsStock]).then(data => {
+     let products = data[0];
+     let productStock = data[1];
+     let joinedProductsInfo = [];
+
+     products.forEach(product => {
+       let productItem = product;
+
+       productItem.count = productStock.filter(
```
Comment (hermandzemski) [inline products-service/services/dynamodb/products-data.service.js]: .find suites better in this case

File: products-service/services/dynamodb/products-data.service.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+         productStockItem => productStockItem.product_id === product.id
+       )[0].count;
```
Comment (hermandzemski) [inline products-service/services/dynamodb/products-data.service.js]: there is an error if there is no record about the particular product in the stock

File: products-service/services/dynamodb/products-data.service.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+export const getAllProducts = async () => {
+   let products = await scanTable(scanProductsParams);
+   let productsStock = await scanTable(scanStocksParams);
+   
+   const productsData = Promise.all([products, productsStock]).then(data => {
+     let products = data[0];
+     let productStock = data[1];
+     let joinedProductsInfo = [];
+
+     products.forEach(product => {
```
Comment (hermandzemski) [inline products-service/services/dynamodb/products-data.service.js]: Can be simplified:

```suggestion
     return products.map(product => {
         const count = productStock.find(p => p.product_id === product.id)?.count ?? 0;
         return { ...product, count: productCount };
     })
```

File: products-service/services/dynamodb/products-data.service.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  } catch (err) {
+    console.log('Error', err);
+  }
+};
+
+export const getAllProducts = async () => {
+   let products = await scanTable(scanProductsParams);
+   let productsStock = await scanTable(scanStocksParams);
+   
+   const productsData = Promise.all([products, productsStock]).then(data => {
```
Comment (hermandzemski) [inline products-service/services/dynamodb/products-data.service.js]: there is no need to use Promise.all as await used above

File: products-service/services/dynamodb/products-data.service.js (hunk 1 above)
Comment (Baravar) [reply products-service/services/dynamodb/products-data.service.js]: yes, I do agree and implement this change - thanks for the advice!

File: products-service/services/dynamodb/products-data.service.js (hunk 2 above)
Comment (Baravar) [reply products-service/services/dynamodb/products-data.service.js]: thanks, going to secure this code

File: products-service/services/dynamodb/products-data.service.js (hunk 3 above)
Comment (Baravar) [reply products-service/services/dynamodb/products-data.service.js]: thanks!

**Comment by Baravar** [comment]
Hi @hermandzemski , all your proposed improvements have been applied and deployed. Thanks

**Pull request merged**